craft_application/models/project.py

```diff
@@ -93,14 +96,114 @@ class BuildPlannerConfig(CraftBaseConfig):
     """The BuildPlanner model uses attributes from the project yaml."""
 
 
+class Platform(CraftBaseModel):
+    """Project platform definition."""
+
+    build_on: list[str] | None = pydantic.Field(min_items=1, unique_items=True)
+    build_for: list[str] | None = pydantic.Field(
+        min_items=1, max_items=1, unique_items=True
+    )
+
+    @pydantic.validator(  # pyright: ignore[reportUnknownMemberType,reportUntypedFunctionDecorator]
+        "build_on", "build_for"
+    )
+    def _validate_architectures(cls, values: list[str]) -> list[str]:
+        """Validate the architecture entries."""
+        for architecture in values:
+            if not is_valid_architecture(architecture):
+                raise errors.CraftValidationError(
+                    f"Invalid architecture: {architecture!r} "
+                    "must be a valid debian architecture."
+                )
+
+        return values
+
+    @pydantic.root_validator(  # pyright: ignore[reportUntypedFunctionDecorator,reportUnknownMemberType]
+        skip_on_failure=True
+    )
+    @classmethod
+    def _validate_platform_set(cls, values: Mapping[str, Any]) -> Mapping[str, Any]:
+        """If build_for is provided, then build_on must also be."""
+        if not values.get("build_on") and values.get("build_for"):
+            raise errors.CraftValidationError(
+                "'build_for' expects 'build_on' to also be provided."
+            )
+
+        return values
+
+
+def _get_default_platform() -> dict[str, Platform]:
+    """Return a platform the builds on and build for the host architecture."""
+    return {
+        get_host_architecture(): Platform(
+            build_on=[get_host_architecture()],
+            build_for=[get_host_architecture()],
+        )
+    }
+
+
+def _populate_platforms(platforms: dict[str, Platform]) -> dict[str, Platform]:
+    """Populate empty platform entries.
+
+    :param platforms: The platform data.
+
+    :returns: The dict of platforms with populated entries.
+    """
+    for platform_label, platform in platforms.items():
+        if not platform:
+            # populate "empty" platforms entries from the platform's name
+            platforms[platform_label] = Platform(
+                build_on=[platform_label], build_for=[platform_label]
+            )
+
+    return platforms
+
+
 class BuildPlanner(CraftBaseModel, metaclass=abc.ABCMeta):
     """The BuildPlanner obtains a build plan for the project."""
 
+    platforms: dict[str, Platform] = _get_default_platform()
```

Comment:
I know that Snapcraft's `platforms` default to the running platform, but Rockcraft doesn't (as far as I know, on purpose). Is it a good idea to have this default here?

Reply:
After an offline discussion, I dropped the default value for `platforms`.